### PR TITLE
feature(multi-az): add possibility to simulate racks

### PR DESCRIPTION
**Pull request by soyacz**
Tests running in multi-az environments generate additional
network-traffic costs. We can simulate racks with
`GossipingPropertyFileSnitch` snitch and provision instances in
one az to avoid them.

Apply `GossipingPropertyFileSnitch` to the nodes and make provisioning
to provision instances only in one az, but configure nodes snitch to
simulate multi-az environment.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/cluster_k8s/__init__.py [hunk 1]
```diff
@@ -2593,10 +2593,6 @@ def connection_bundle_file(self) -> Path | None:
     def node_setup(self, node: BaseScyllaPodContainer, verbose: bool = False, timeout: int = 3600):
         if self.test_config.BACKTRACE_DECODING:
             node.install_scylla_debuginfo()
```
Comment (vponomaryov) [inline sdcm/cluster_k8s/__init__.py:2595]: This file change looks like a `must be a separate PR`.

File: sdcm/sct_config.py [hunk 2]
```diff
@@ -1785,6 +1789,10 @@ def __init__(self):
                     raise ValueError(
                         f" Got error: {repr(exp)}, on item '{param}'") from exp
 
+        # 15 Force endpoint_snitch to GossipingPropertyFileSnitch if using simulated_racks
+        if self.get("simulated_racks") > 1:
+            self["endpoint_snitch"] = "org.apache.cassandra.locator.GossipingPropertyFileSnitch"
```
Comment (vponomaryov) [inline sdcm/sct_config.py:1794]: This `silent` overwriting looks dangerous to me.
Maybe we should check the value of this option and if it is already set then fail the test.
If it is not set then define it like here.

File: sdcm/snitch_configuration.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        requires_restart = False
+        properties = {name: str(getattr(self, name)).lower() for name in dir(self)
+                      if isinstance(getattr(self.__class__, name, None), property)}
+        rackdc_content = ""
+        for key, value in properties.items():
+            if value:
+                rackdc_content += f"{key}={str(value).lower()}\n"
+        if rackdc_content:
+            requires_restart = True
+            cmd = f"sudo sh -c 'echo \"{rackdc_content}\" > {SCYLLA_PROPERTIES_PATH}'"
```
Comment (vponomaryov) [inline sdcm/snitch_configuration.py:64]: It will fail on K8S.

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/cluster_k8s/__init__.py [hunk 4]
```diff
@@ -2593,10 +2593,6 @@ def connection_bundle_file(self) -> Path | None:
...
-
-        if self.test_config.MULTI_REGION:
-            node.datacenter_setup(self.datacenter)  # pylint: disable=no-member
```
Comment (fruch) [inline sdcm/cluster_k8s/__init__.py:2598]: removing this might be problematic, for non AWS cases.
we should test some GCP multi-region case to validate.

File: sdcm/sct_config.py (hunk 2 above)
Comment (fruch) [reply sdcm/sct_config.py:1794]: I think failing on the configuration level in this case seems o.k. to me, and should be enough

File: sdcm/snitch_configuration.py (hunk 3 above)
Comment (fruch) [reply sdcm/snitch_configuration.py:64]: cause of the sudo ?

File: sdcm/cluster_k8s/__init__.py (hunk 4 above)
Comment (vponomaryov) [reply sdcm/cluster_k8s/__init__.py:2598]: This is `k8s` file.

File: sdcm/snitch_configuration.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            self._node.remoter.run(cmd)
```
Comment (fruch) [inline sdcm/snitch_configuration.py:65]:
```suggestion
            self._node.remoter.sudo(cmd)
```

File: sdcm/snitch_configuration.py (hunk 3 above)
Comment (vponomaryov) [reply sdcm/snitch_configuration.py:64]: Because we handle rackds and scylla yaml files via the configmap objects.

Comment (fruch) [reply sdcm/snitch_configuration.py:64]: so we should implement `remote_cassandra_rackdc_properties` for VMs ? to align ?

File: sdcm/sct_config.py (hunk 2 above)
Comment (soyacz) [reply sdcm/sct_config.py:1794]: ok

File: sdcm/snitch_configuration.py (hunk 5 above)
Comment (soyacz) [reply sdcm/snitch_configuration.py]: no longer needed after using `remote_cassandra_rackdc_properties`

File: sdcm/cluster_k8s/__init__.py (hunk 1 above)
Comment (soyacz) [reply sdcm/cluster_k8s/__init__.py:2595]: it is...

File: sdcm/snitch_configuration.py (hunk 3 above)
Comment (soyacz) [reply sdcm/snitch_configuration.py]: `remote_cassandra_rackdc_properties` is now used

**Comment by soyacz** [comment]
effectively, racks look as follows:
```
< t:2023-05-10 11:06:00,326 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > Datacenter: eu-westscylla_node_west
< t:2023-05-10 11:06:00,394 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > ===================================
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > Status=Up/Down
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > |/ State=Normal/Leaving/Joining/Moving
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > --  Address      Load       Tokens       Owns    Host ID                               Rack
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.2.183   ?          256          ?       5352042a-4944-4567-a168-ce0e1ee50f85  rack2
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.64    5.21 MB    256          ?       1d2f873d-df7f-4606-bd84-e7b4ae8182f5  rack1
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.0.28    9.41 MB    256          ?       592167bf-f97e-4182-8184-445e1e701bc6  rack0
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > Datacenter: us-eastscylla_node_east
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > ===================================
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > Status=Up/Down
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > |/ State=Normal/Leaving/Joining/Moving
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > --  Address      Load       Tokens       Owns    Host ID                               Rack
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.12.0.187  ?          256          ?       07e30f9d-1d3d-499f-b5ac-1bcc58075603  rack0
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.12.1.180  ?          256          ?       bb6d4ca0-e580-4d64-9baa-3b4d1ba99816  rack1
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.12.3.164  7.86 MB    256          ?       f2c54815-c077-410f-88e9-7ebd8e0e3b36  rack0
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.12.1.102  8.83 MB    256          ?       25ccf691-7443-40d7-934c-2c5b458d6a60  rack2
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.12.3.231  6.44 MB    256          ?       66613d06-aaa4-4847-86ae-3e2db3c7ef7a  rack2
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.12.1.81   8.1 MB     256          ?       8f5e5a1f-aa3f-4222-bae5-18836ebf93a4  rack1
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > Datacenter: us-west-2scylla_node_west
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > =====================================
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > Status=Up/Down
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > |/ State=Normal/Leaving/Joining/Moving
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > --  Address      Load       Tokens       Owns    Host ID                               Rack
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.15.3.14   ?          256          ?       ab0139e6-1c23-437e-a3b8-07cc9754380b  rack0
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.15.1.215  9.29 MB    256          ?       531036e8-30c3-4c5d-a279-f979c32d5db5  rack2
< t:2023-05-10 11:06:00,395 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.15.0.19   ?          256          ?       4eafeb12-73c2-47ed-9434-7deaf0249291  rack1
```

**Comment by soyacz** [comment]
I rebased and added eks provision test to verify if doesn't break there anything.

**Comment by soyacz** [comment]
@vponomaryov @fruch please re-review

**Review by fruch — DISMISSED** [review]
LGTM

File: sdcm/cluster_k8s/__init__.py (hunk 1 above)
Comment (vponomaryov) [reply sdcm/cluster_k8s/__init__.py:2663]: It is still here...

Comment (soyacz) [reply sdcm/cluster_k8s/__init__.py:2663]: true, I read it as `separate commit` - now I see you meant `PR`

File: jenkins-pipelines/longevity-cdc-100gb-4h.jenkinsfile [hunk 6]
```diff
@@ -6,7 +6,7 @@ def lib = library identifier: 'sct@snapshot', retriever: legacySCM(scm)
 longevityPipeline(
     backend: 'aws',
     region: 'eu-west-1',
-    availability_zone: 'a,b,c',
+    availability_zone: 'a',
```
[the remark this answers is not in the document]
Comment (soyacz) [reply jenkins-pipelines/longevity-cdc-100gb-4h.jenkinsfile:9]: no, it's enough to place it into test config. az setting was different as was overridden by jenkins pipeline setting.

**Review by vponomaryov — APPROVED** [review]

**Review by fruch — APPROVED** [review]
LGTM

**Comment by roydahan** [comment]
@fruch / @soyacz let's present it in weekly meeting so people will be aware of it.
Showing which longevities are "simulated racks" and which are real "AZ" and how to differ (quite easy but make sure people are aware of it).
